Added Animated Hamburger Menu

sushantgwr87 opened the pull request:
# Animated Hamburger Menu Pull Request

## What does this PR do?
Fixes #245
[Link to the issue](https://github.com/meilisearch/landing/issues/245)

## Description
Added styled component with animation for mobile header burger menu. Removed svg icons used for it as well.

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

bidoubiwa commented:
Hey @sushantgwr87 thanks so much for you contribution 🙏
Could you fix the lining issues ? Its hard to read the code between the errors.

```
yarn lint
```

sushantgwr87 commented:
@bidoubiwa Updated the Code and Removed Errors. Kindly Check Once Again.

bidoubiwa commented:
I still see linting errors 😔 See https://github.com/meilisearch/landing/pull/488/files

sushantgwr87 commented:
Yeah, Actually my system shows different lint errors and here bit more. No problem I will fix all (this time both check) and will let you know.

sushantgwr87 commented:
Sir Kindly run the checks again, this time I guess errors should be gone. Before I was not using prettier extension that's why it linted different errors in VS code.

bidoubiwa reviewed (changes requested):
✨

bidoubiwa commented on src/blocks/Header/MobileHeader.js:
```diff
@@ -71,8 +100,12 @@ const MobileHeader = ({ headerProps }) => {
       <Logo />
-      <DialogDisclosure {...dialog} style={{ color: 'white', height: 24 }}>
-        {dialog.visible ? <Cross height={24} /> : <Hamburger height={24} />}
+      <DialogDisclosure {...dialog} style={{ color: 'white', height: '24px' }}>
+        <MenuBtnWrapper clicked={dialog.visible}>
+          <MenuBtnBar />
+          <MenuBtnBar />
+          <MenuBtnBar />
+        </MenuBtnWrapper>
```
Could you create a `Hamburger` component out of it and add it in an apart file?

sushantgwr87 replied:
Sure. working on it. I will create file in same folder regarding it.

bidoubiwa replied:
I prefered you add it in the `components` folder.

sushantgwr87 commented:
I made it. Check once if its correct or not.

I created Hamburger Taking whole DialogDisclosure part in it and then importing the hamburger in MobileHeader.js.

bidoubiwa reviewed (commented):
You can leave DialogDisclosure in MobileHeader.js

sushantgwr87 commented:
> You can leave DialogDisclosure in MobileHeader.js

Okay Updating it.

bidoubiwa reviewed (changes requested):
Could you add the Hamburger in the components folder ?

sushantgwr87 commented:
> Could you add the Hamburger in the components folder ?

Yes Updating it

sushantgwr87 commented:
Updated Sir, Check Once.

bidoubiwa commented:
Hey @sushantgwr87 

Since @mdubus is not present, I can not be sure that everything is as wanted. So unfortunately, we'll have to wait until next week to approve or not your pr. 

These are the questions that I'm thinking about:

- The stories should be updated
- The unused icons should be removed
- Should the animation be done with `lottie` as mentioned in the issue?
- In what folder should the component be ? As maybe I'm wrong

sushantgwr87 commented:
> Hey @sushantgwr87
> 
> Since @mdubus is not present, I can not be sure that everything is as wanted. So unfortunately, we'll have to wait until next week to approve or not your pr.
> 
> These are the questions that I'm thinking about:
> 
> * The stories should be updated
> * The unused icons should be removed
> * Should the animation be done with `lottie` as mentioned in the issue?
> * In what folder should the component be ? As maybe I'm wrong

No Problem. Will See it on next Week.

sushantgwr87 commented:
> Hello @sushantgwr87 👋
> 
> Thanks a lot for your PR and sorry for the inconvenience of me not being there during the first week of Hacktoberfest.
> 
> Is there a reason why you chose to create a custom Component for this instead of using Lottie as recommended in the issue?
> 
> Thanks a lot and have a good day ✨

@mdubus No Sir. There is no specific reason for not using Lottie file for this. Actually we are already using styled components so I thought this animation can be easily created without much fuss and with little bit of css animation, it will work as needed. 

Though If you think I should use Lottie file for it, I can work on that and create it using that.

mdubus commented:
> Though If you think I should use Lottie file for it, I can work on that and create it using that.

Thanks a lot for your answer 🙏 Indeed it would be great if you could use Lottie instead, for multiple reasons:

- I see a little CSS problem inside your animation, which would be fixed with Lottie
- It will be easier to maintain as every animation will be in the same format and in the same place
- It will require less code

Thanks a lot for your time, and I deeply apologize for the inconvenience 🙏

sushantgwr87 commented:
> > Though If you think I should use Lottie file for it, I can work on that and create it using that.
> 
> Thanks a lot for your answer 🙏 Indeed it would be great if you could use Lottie instead, for multiple reasons:
> 
> * I see a little CSS problem inside your animation, which would be fixed with Lottie
> * It will be easier to maintain as every animation will be in the same format and in the same place
> * It will require less code
> 
> Thanks a lot for your time, and I deeply apologize for the inconvenience 🙏

Understood and No Worries. I will ping you once I fix it. Thank you.

sushantgwr87 commented:
Hello,

A little help/discussion needed regarding this problem.

**Problem**: Complete Animation runs on button click.

I created the Lottie file and used already made Lottie Component to create the Hamburger menu. But problem is I am not able to control the animation like to work accordingly with toggle.

I researched about it and found that segments are used for this purpose but the `react-lottie-wrapper` does not work or has an issue regarding it. Details of issue [here](https://github.com/felippenardi/lottie-react-web/issues/30). And I don't think adding one more lottie package is good.

I am trying to find other ways but any insight regarding this will help immensely.

Lottie Code:
```
<Lottie
      animation={hamburgerMenuAnimation}
      options={{
            loop: false,
            autoplay: false,
      }}
      segments={[0,50]}
      ariaLabel="Hamburger Menu"
/>
```

Lottie file used (white version of this):
https://lottiefiles.com/9789-burger-menu

Video for Problem:
https://user-images.githubusercontent.com/47962312/195096090-56516901-f152-4eb7-b0f4-4342de6d25d3.mp4

sushantgwr87 commented:
Hello,

Need help in fixing Lottie component or code for animation.

I created Hamburger menu animation using Lottie file and already made Lottie component. Problem is whole animation runs on button click from = to x back to =. (Lottie file has same animation too)

I tried using segments but seems they don't work with this package I guess as it does not shows props relating it or such toggle support. I am still researching on this for solution but any insights will help surely.

Lottie Code:
```
<Lottie
          animation={hamburgerMenuAnimation}
          options={{
            loop: false,
            autoplay: false,
          }}
          ariaLabel="Hamburger Menu"
/>
```

Video for problem:
https://user-images.githubusercontent.com/47962312/195103587-c5935543-3673-4f10-a6bc-c48be563200d.mp4

mdubus commented:
Hello @sushantgwr87 👋 

Thanks a lot for reporting here your issues with the Lottie file. Indeed, I can see can the Lottie animation I provided in the issue goes from front to back and then back to front, and it's not ideal in our context. 
I saw on Lottie's documentation that they have a `segment` option that could allow us to select a section to play (we could have chosen only the beginning of the animation), but unfortunately, it is not yet implemented in the react-lottie library. 

Therefore, I adapted the Lottie animation thanks to the [Lottie Editor](edit.lottiefiles.com), and I just cropped the end of the animation. Here it is: 
[hamburger-menu.json.zip](https://github.com/meilisearch/landing/files/9772376/hamburger-menu.json.zip)

(as GitHub doesn't support json files as attachments, I just compressed it as a zip. You'll have to unzip it 😉). 

Then, you can play it from back to front regarding if the modal is opened or not, with the direction option: `direction={dialog.visible ? 1 : -1}`

Don't hesitate if you have any questions or if you need further help on this. 

Have a nice day! ✨

sushantgwr87 commented:
Hello Ma'am :smiley:,

Great Thanks for that Lottie animation file and guidance as well.

I updated the Lottie code in MobileHeader File:
[Link to file code](https://github.com/sushantgwr87/landing/blob/sushantgwr87-Animated-Burger-fix/src/blocks/Header/MobileHeader.js)

```
<Lottie
          direction={dialog.visible ? 1 : -1}
          animation={hamburgerMenuAnimation}
          options={{
            loop: false,
            autoplay: false,
          }}
          ariaLabel="Hamburger Menu"
 />
```

I added your given Lottie json file in Lottie folder and imported them as needed.

Kindly check once if everything is correct or any other update/fix is needed.

Thanks for help :smile_cat:

mdubus commented on src/blocks/Header/MobileHeader.js:
```diff
@@ -68,11 +69,49 @@ const MobileHeader = ({ headerProps }) => {
     unstable_initialFocusRef: null,
   })
+
+  const [sequence, setSequence] = useState({
+    segments: [0, 1],
+    forceFlag: true,
+  })
+
+  const animate = () => {
+    console.log('run')
+    console.log(sequence)
+    let start = sequence.segments[0]
+    let stop = sequence.segments[1]
+
+    if (stop === 1 || stop === 90) {
+      start = 0
+      stop = 50
+    } else {
+      start = 50
+      stop = 75
+    }
+
+    setSequence({
+      segments: [start, stop],
+      forceFlag: true,
+    })
+  }
+
```
I don't think that this is necessary. Can you please check if everything works without it? 🙏

sushantgwr87 replied:
Sorry, I forgot to remove segment function which I created. Fixing it.

sushantgwr87 replied:
Fixed it

mdubus reviewed (changes requested):
Almost perfect ✨

Can you please:
- Check if the old `Hamburger` and `Cross` icon are still used, and delete them if it's not the case (check the svg icon & rerun the `yarn icons` command if necessary, also check the `Icons` story of Storybook) 🙏
- Rebase your PR on main

Thanks a lot! 🦕

sushantgwr87 commented:
Hello,

Sorry for mess-up and for any inconvenience I caused during rebasing and updating files.

Actually I am still confused what should I do regarding `rebasing my PR to main`. I have never done this so I am clueless, If you can guide me then it will very helpful.

I have updated both main and this repo, after messing things up badly :smiling_face_with_tear: . I used rebasing commands to update branches but still confused what actually I need to do here.

#### Other Details:
I have removed unused Icons from MobileHeader.js File and checked for any other icons. I have not deleted any icon file though.

Thanks and Sorry for asking too much help :sweat:

mdubus commented:
Hello @sushantgwr87  

No problem 😀 
I can see some conflicts in your PR that weren't quite well resolved and would erase some previous work. 
Since you are not comfortable with git, I would suggest starting fresh and creating a brand new PR on a brand new branch 🚀

Here is how you can achieve it:
- Go back to your main branch 
- Pull the changes that were made on the `main` branch from the Meilisearch repository
- Create a new branch
- Add the changes you want to the code
- Then push the branch and open a new PR

Hope this helps ✨

sushantgwr87 commented:
Thanks for guide Ma'am.

I updated my main repo and opened a new open request.

You can check here [PR](https://github.com/meilisearch/landing/pull/513).

mdubus commented:
Thanks @sushantgwr87 🙏

Closing in favor of #513
